jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java

```diff
@@ -471,6 +511,144 @@ default boolean rewind()
      */
     public interface Sink
     {
+        /**
+         * <p>Wraps the given {@link OutputStream} as a {@link Sink}.
+         * @param out The stream to wrap
+         * @return A sink wrapping the stream
+         */
+        static Sink from(OutputStream out)
+        {
+            return new Sink()
+            {
+                boolean closed;
+
+                @Override
+                public void write(boolean last, ByteBuffer byteBuffer, Callback callback)
+                {
+                    if (closed)
+                    {
+                        callback.failed(new EOFException());
+                        return;
+                    }
+                    try
+                    {
+                        BufferUtil.writeTo(byteBuffer, out);
+                        if (last)
+                        {
+                            closed = true;
+                            out.close();
+                        }
+                        callback.succeeded();
+                    }
+                    catch (Throwable t)
+                    {
+                        callback.failed(t);
+                    }
+                }
+            };
+        }
+
+        /**
+         * <p>Wraps the given {@link ByteChannel} as a {@link Sink}.
+         * @param channel The {@link ByteChannel} to wrap
+         * @return A sink wrapping the stream
+         */
+        static Sink from(ByteChannel channel)
+        {
+            return new Sink()
+            {
+                boolean closed;
+
+                @Override
+                public void write(boolean last, ByteBuffer byteBuffer, Callback callback)
+                {
+                    if (closed)
+                    {
+                        callback.failed(new EOFException());
+                        return;
+                    }
+                    try
+                    {
+                        int remaining = byteBuffer.remaining();
+                        int tries = 0;
```

Comment:
This may well be legacy and that we now never get spurious failures to write all the bytes... but that is hard to test for all JVMs on all OS's.  So I don't think we can remove such loops.

However, I think this implementation could be improved and we can keep looping whilst progress is being made.